app/post_metadata.go

```diff
@@ -386,15 +395,55 @@ func getLinkMetadataFromCache(requestURL string) (*opengraph.OpenGraph, *model.P
 	}
 }
 
-func cacheLinkMetadata(requestURL string, og *opengraph.OpenGraph, image *model.PostImage) {
+func (a *App) getLinkMetadataFromDatabase(requestURL string, timestamp int64) (*opengraph.OpenGraph, *model.PostImage, bool) {
+	result := <-a.Srv.Store.LinkMetadata().Get(requestURL, timestamp)
+	if result.Err != nil {
+		return nil, nil, false
+	}
+
+	data := result.Data.(*model.LinkMetadata).Data
+
+	switch v := data.(type) {
+	case *opengraph.OpenGraph:
+		return v, nil, true
+	case *model.PostImage:
+		return nil, v, true
+	default:
+		return nil, nil, true
+	}
+}
+
+func (a *App) saveLinkMetadataToDatabase(requestURL string, timestamp int64, og *opengraph.OpenGraph, image *model.PostImage) {
+	metadata := &model.LinkMetadata{
+		URL:       requestURL,
+		Timestamp: timestamp,
+	}
+
+	if og != nil {
+		metadata.Type = model.LINK_METADATA_TYPE_OPENGRAPH
+		metadata.Data = og
+	} else if image != nil {
+		metadata.Type = model.LINK_METADATA_TYPE_IMAGE
+		metadata.Data = image
+	} else {
+		metadata.Type = model.LINK_METADATA_TYPE_NONE
+	}
+
+	result := <-a.Srv.Store.LinkMetadata().Save(metadata)
+	if result.Err != nil {
+		mlog.Warn("Failed to write link metadata", mlog.String("request_url", requestURL), mlog.Err(result.Err))
```

Comment:
It seems more idiomatic to return this error and allow the caller to decide if it's going to simply log it.

Reply:
The reason that I chose to just log it here is that it's more of a warning since we were able to actually get the metadata fine. If we were to pass it further, it would just be `App.getLinkMetadata` that's logging it since passing it any further would cause the metadata to not even be returned as the code that calls `getLinkMetadata` would not be able to understand that this is just a warning and not a real error.

If this were something that was going to be reused, I'd lean more towards passing the error as far as possible, but I don't think it makes a difference when this is a very specific method that won't be used elsewhere.